File: android-native/app/src/main/kotlin/com/wordwheel/game/Level.kt
package com.wordwheel.game

enum class Dir { ACROSS, DOWN }

data class PlacedWord(
    val word: String,
    val row: Int,
    val col: Int,
    val dir: Dir,
)

data class Level(
    val rows: Int,
    val cols: Int,
    val letters: List<Char>,
    val words: List<PlacedWord>,
) {
    fun answersSet(): Set<String> = words.map { it.word }.toSet()

    fun usedCells(): Set<Pair<Int, Int>> {
        val used = mutableSetOf<Pair<Int, Int>>()
        for (pw in words) {
            for (i in pw.word.indices) {
                val (r, c) = when (pw.dir) {
                    Dir.ACROSS -> pw.row to (pw.col + i)
                    Dir.DOWN -> (pw.row + i) to pw.col
                }
                used.add(r to c)
            }
        }
        return used
    }

    fun solutionLetterAt(row: Int, col: Int): Char? {
        for (pw in words) {
            for ((i, ch) in pw.word.withIndex()) {
                val (r, c) = when (pw.dir) {
                    Dir.ACROSS -> pw.row to (pw.col + i)
                    Dir.DOWN -> (pw.row + i) to pw.col
                }
                if (r == row && c == col) return ch
            }
        }
        return null
    }

    fun bonusWords(): List<String> = when (letters) {
        listOf('C', 'A', 'T', 'S') -> listOf("ACT", "ACTS", "SCAT", "CATS")
        listOf('S', 'P', 'I', 'N', 'E') -> listOf("PIN", "PIE", "SIP", "SNIP", "PIES", "PENS", "SINE", "PINE")
        listOf('H', 'A', 'S', 'T', 'E') -> listOf("HEAT", "SET", "HAT", "TEA", "ATE", "ETA", "EAST", "EATS")
        listOf('W', 'A', 'R', 'M', 'S') -> listOf("WARMS", "WARS", "ARMS", "MARS", "ARM", "MAR", "RAW", "WAR")
        listOf('C', 'A', 'R', 'E', 'S') -> listOf("RACE", "CARS", "EARS", "ACE", "ARC", "SCAR", "ARCS", "ACRE", "ACES")
        listOf('G', 'R', 'I', 'N', 'D', 'S') -> listOf("GRINDS", "GRID", "RIND", "DING", "RING", "GRIN", "RINGS")
        listOf('P', 'L', 'A', 'N', 'E', 'T', 'S') -> listOf("PLAN", "PLANT", "LEAN", "PLATE", "STEAL", "LANE", "TALE", "PELT")
        listOf('C', 'R', 'A', 'N', 'E', 'S', 'D') -> listOf("DANCE", "SCARE", "RACED", "CEDAR", "CANE", "ACRE", "RAN")
        listOf('S', 'T', 'O', 'R', 'E', 'D') -> listOf("RODE", "DOER", "STORE", "DOTES", "TROD", "ODES", "SORE", "TORE")
        listOf('T', 'R', 'A', 'I', 'N', 'S', 'E') -> listOf("STARE", "RETAIN", "SATIRE", "INSERT", "STAIN", "RAIN", "RISE", "TIRE")
        else -> emptyList()
    }

    companion object {
        const val TOTAL_LEVELS = 10

        fun get(levelNum: Int): Level = when (levelNum) {
            1 -> Level(
                rows = 5, cols = 5,
                letters = listOf('C', 'A', 'T', 'S'),
                words = listOf(
                    PlacedWord("CAST", 1, 0, Dir.ACROSS),
                    PlacedWord("CAT", 1, 0, Dir.DOWN),
                    PlacedWord("SAT", 1, 2, Dir.DOWN),
                ),
            )
            2 -> Level(
                rows = 6, cols = 6,
                letters = listOf('S', 'P', 'I', 'N', 'E'),
                words = listOf(
                    PlacedWord("SPINE", 1, 0, Dir.ACROSS),
                    PlacedWord("SIN", 1, 0, Dir.DOWN),
                    PlacedWord("NIP", 1, 3, Dir.DOWN),
                    PlacedWord("PEN", 3, 3, Dir.ACROSS),
                ),
            )
            3 -> Level(
                rows = 7, cols = 7,
                letters = listOf('H', 'A', 'S', 'T', 'E'),
                words = listOf(
                    PlacedWord("HASTE", 1, 0, Dir.ACROSS),
                    PlacedWord("HATE", 1, 0, Dir.DOWN),
                    PlacedWord("SEAT", 1, 2, Dir.DOWN),
                    PlacedWord("EAT", 4, 0, Dir.ACROSS),
                ),
            )
            4 -> Level(
                rows = 7, cols = 7,
                letters = listOf('W', 'A', 'R', 'M', 'S'),
                words = listOf(
                    PlacedWord("SWARM", 1, 0, Dir.ACROSS),
                    PlacedWord("SAW", 1, 0, Dir.DOWN),
                    PlacedWord("RAM", 1, 3, Dir.DOWN),
                    PlacedWord("WARM", 3, 0, Dir.ACROSS),
                ),
            )
            5 -> Level(
                rows = 7, cols = 7,
                letters = listOf('C', 'A', 'R', 'E', 'S'),
                words = listOf(
                    PlacedWord("CARES", 1, 0, Dir.ACROSS),
                    PlacedWord("CARE", 1, 0, Dir.DOWN),
                    PlacedWord("SEAR", 1, 4, Dir.DOWN),
                    PlacedWord("ERA", 4, 0, Dir.ACROSS),
                ),
            )
            6 -> Level(
                rows = 7, cols = 7,
                letters = listOf('G', 'R', 'I', 'N', 'D', 'S'),
                words = listOf(
                    PlacedWord("GRINS", 1, 0, Dir.ACROSS),
                    PlacedWord("GIN", 1, 0, Dir.DOWN),
                    PlacedWord("SIR", 1, 4, Dir.DOWN),
                    PlacedWord("RID", 3, 4, Dir.ACROSS),
                    PlacedWord("DIG", 3, 6, Dir.DOWN),
                ),
            )
            7 -> Level(
                rows = 8, cols = 8,
                letters = listOf('P', 'L', 'A', 'N', 'E', 'T', 'S'),
                words = listOf(
                    PlacedWord("PLANETS", 1, 0, Dir.ACROSS),
                    PlacedWord("PETAL", 1, 0, Dir.DOWN),
                    PlacedWord("NETS", 1, 3, Dir.DOWN),
                    PlacedWord("SLANT", 1, 6, Dir.DOWN),
                    PlacedWord("ANTS", 4, 0, Dir.ACROSS),
                ),
            )
            8 -> Level(
                rows = 8, cols = 8,
                letters = listOf('C', 'R', 'A', 'N', 'E', 'S', 'D'),
                words = listOf(
                    PlacedWord("CRANES", 1, 0, Dir.ACROSS),
                    PlacedWord("CARED", 1, 0, Dir.DOWN),
                    PlacedWord("NEARS", 1, 3, Dir.DOWN),
                    PlacedWord("SANE", 1, 5, Dir.DOWN),
                    PlacedWord("DENS", 5, 0, Dir.ACROSS),
                    PlacedWord("END", 5, 1, Dir.DOWN),
                ),
            )
            9 -> Level(
                rows = 8, cols = 8,
                letters = listOf('S', 'T', 'O', 'R', 'E', 'D'),
                words = listOf(
                    PlacedWord("STORED", 1, 0, Dir.ACROSS),
                    PlacedWord("SORT", 1, 0, Dir.DOWN),
                    PlacedWord("REST", 1, 3, Dir.DOWN),
                    PlacedWord("DOSE", 1, 5, Dir.DOWN),
                    PlacedWord("ROES", 3, 0, Dir.ACROSS),
                ),
            )
            else -> Level(
                rows = 9, cols = 9,
                letters = listOf('T', 'R', 'A', 'I', 'N', 'S', 'E'),
                words = listOf(
                    PlacedWord("TRAINS", 1, 0, Dir.ACROSS),
                    PlacedWord("TEARS", 1, 0, Dir.DOWN),
                    PlacedWord("INSET", 1, 3, Dir.DOWN),
                    PlacedWord("SIREN", 1, 5, Dir.DOWN),
                    PlacedWord("ANTS", 3, 0, Dir.ACROSS),
                    PlacedWord("SENT", 5, 0, Dir.ACROSS),
                ),
            )
        }
    }
}

File: ios-native/WordWheel/Models/Level.swift
import Foundation

enum Dir {
    case across
    case down
}

struct PlacedWord {
    let word: String
    let row: Int
    let col: Int
    let dir: Dir
}

struct Cell: Hashable {
    let row: Int
    let col: Int
}

struct Level {
    static let totalLevels = 10

    let rows: Int
    let cols: Int
    let letters: [Character]
    let words: [PlacedWord]

    var answersSet: Set<String> { Set(words.map { $0.word }) }

    var usedCells: Set<Cell> {
        var used = Set<Cell>()
        for pw in words {
            for i in 0..<pw.word.count {
                let cell: Cell
                switch pw.dir {
                case .across: cell = Cell(row: pw.row, col: pw.col + i)
                case .down:   cell = Cell(row: pw.row + i, col: pw.col)
                }
                used.insert(cell)
            }
        }
        return used
    }

    func solutionLetter(row: Int, col: Int) -> Character? {
        for pw in words {
            for (i, ch) in pw.word.enumerated() {
                let r: Int, c: Int
                switch pw.dir {
                case .across: r = pw.row; c = pw.col + i
                case .down:   r = pw.row + i; c = pw.col
                }
                if r == row && c == col { return ch }
            }
        }
        return nil
    }

    var bonusWords: [String] {
        switch String(letters) {
        case "CATS":     return ["ACT", "ACTS", "SCAT", "CATS"]
        case "SPINE":    return ["PIN", "PIE", "SIP", "SNIP", "PIES", "PENS", "SINE", "PINE"]
        case "HASTE":    return ["HEAT", "SET", "HAT", "TEA", "ATE", "ETA", "EAST", "EATS"]
        case "WARMS":    return ["WARMS", "WARS", "ARMS", "MARS", "ARM", "MAR", "RAW", "WAR"]
        case "CARES":    return ["RACE", "CARS", "EARS", "ACE", "ARC", "SCAR", "ARCS", "ACRE", "ACES"]
        case "GRINDS":   return ["GRINDS", "GRID", "RIND", "DING", "RING", "GRIN", "RINGS"]
        case "PLANETS":  return ["PLAN", "PLANT", "LEAN", "PLATE", "STEAL", "LANE", "TALE", "PELT"]
        case "CRANESD":  return ["DANCE", "SCARE", "RACED", "CEDAR", "CANE", "ACRE", "RAN"]
        case "STORED":   return ["RODE", "DOER", "STORE", "DOTES", "TROD", "ODES", "SORE", "TORE"]
        case "TRAINSE":  return ["STARE", "RETAIN", "SATIRE", "INSERT", "STAIN", "RAIN", "RISE", "TIRE"]
        default:         return []
        }
    }

    static func get(_ levelNum: Int) -> Level {
        switch levelNum {
        case 1:
            return Level(rows: 5, cols: 5, letters: ["C","A","T","S"], words: [
                PlacedWord(word: "CAST", row: 1, col: 0, dir: .across),
                PlacedWord(word: "CAT",  row: 1, col: 0, dir: .down),
                PlacedWord(word: "SAT",  row: 1, col: 2, dir: .down),
            ])
        case 2:
            return Level(rows: 6, cols: 6, letters: ["S","P","I","N","E"], words: [
                PlacedWord(word: "SPINE", row: 1, col: 0, dir: .across),
                PlacedWord(word: "SIN",   row: 1, col: 0, dir: .down),
                PlacedWord(word: "NIP",   row: 1, col: 3, dir: .down),
                PlacedWord(word: "PEN",   row: 3, col: 3, dir: .across),
            ])
        case 3:
            return Level(rows: 7, cols: 7, letters: ["H","A","S","T","E"], words: [
                PlacedWord(word: "HASTE", row: 1, col: 0, dir: .across),
                PlacedWord(word: "HATE",  row: 1, col: 0, dir: .down),
                PlacedWord(word: "SEAT",  row: 1, col: 2, dir: .down),
                PlacedWord(word: "EAT",   row: 4, col: 0, dir: .across),
            ])
        case 4:
            return Level(rows: 7, cols: 7, letters: ["W","A","R","M","S"], words: [
                PlacedWord(word: "SWARM", row: 1, col: 0, dir: .across),
                PlacedWord(word: "SAW",   row: 1, col: 0, dir: .down),
                PlacedWord(word: "RAM",   row: 1, col: 3, dir: .down),
                PlacedWord(word: "WARM",  row: 3, col: 0, dir: .across),
            ])
        case 5:
            return Level(rows: 7, cols: 7, letters: ["C","A","R","E","S"], words: [
                PlacedWord(word: "CARES", row: 1, col: 0, dir: .across),
                PlacedWord(word: "CARE",  row: 1, col: 0, dir: .down),
                PlacedWord(word: "SEAR",  row: 1, col: 4, dir: .down),
                PlacedWord(word: "ERA",   row: 4, col: 0, dir: .across),
            ])
        case 6:
            return Level(rows: 7, cols: 7, letters: ["G","R","I","N","D","S"], words: [
                PlacedWord(word: "GRINS", row: 1, col: 0, dir: .across),
                PlacedWord(word: "GIN",   row: 1, col: 0, dir: .down),
                PlacedWord(word: "SIR",   row: 1, col: 4, dir: .down),
                PlacedWord(word: "RID",   row: 3, col: 4, dir: .across),
                PlacedWord(word: "DIG",   row: 3, col: 6, dir: .down),
            ])
        case 7:
            return Level(rows: 8, cols: 8, letters: ["P","L","A","N","E","T","S"], words: [
                PlacedWord(word: "PLANETS", row: 1, col: 0, dir: .across),
                PlacedWord(word: "PETAL",   row: 1, col: 0, dir: .down),
                PlacedWord(word: "NETS",    row: 1, col: 3, dir: .down),
                PlacedWord(word: "SLANT",   row: 1, col: 6, dir: .down),
                PlacedWord(word: "ANTS",    row: 4, col: 0, dir: .across),
            ])
        case 8:
            return Level(rows: 8, cols: 8, letters: ["C","R","A","N","E","S","D"], words: [
                PlacedWord(word: "CRANES", row: 1, col: 0, dir: .across),
                PlacedWord(word: "CARED",  row: 1, col: 0, dir: .down),
                PlacedWord(word: "NEARS",  row: 1, col: 3, dir: .down),
                PlacedWord(word: "SANE",   row: 1, col: 5, dir: .down),
                PlacedWord(word: "DENS",   row: 5, col: 0, dir: .across),
                PlacedWord(word: "END",    row: 5, col: 1, dir: .down),
            ])
        case 9:
            return Level(rows: 8, cols: 8, letters: ["S","T","O","R","E","D"], words: [
                PlacedWord(word: "STORED", row: 1, col: 0, dir: .across),
                PlacedWord(word: "SORT",   row: 1, col: 0, dir: .down),
                PlacedWord(word: "REST",   row: 1, col: 3, dir: .down),
                PlacedWord(word: "DOSE",   row: 1, col: 5, dir: .down),
                PlacedWord(word: "ROES",   row: 3, col: 0, dir: .across),
            ])
        default:
            return Level(rows: 9, cols: 9, letters: ["T","R","A","I","N","S","E"], words: [
                PlacedWord(word: "TRAINS", row: 1, col: 0, dir: .across),
                PlacedWord(word: "TEARS",  row: 1, col: 0, dir: .down),
                PlacedWord(word: "INSET",  row: 1, col: 3, dir: .down),
                PlacedWord(word: "SIREN",  row: 1, col: 5, dir: .down),
                PlacedWord(word: "ANTS",   row: 3, col: 0, dir: .across),
                PlacedWord(word: "SENT",   row: 5, col: 0, dir: .across),
            ])
        }
    }
}
